appinventor/buildserver/src/com/google/appinventor/buildserver/AabCompiler.java

```diff
@@ -0,0 +1,314 @@
+
+  private static File createDir(File parentDir, String name) {
+    File dir = new File(parentDir, name);
+    if (!dir.exists()) {
+      dir.mkdir();
+    }
+    return dir;
+  }
+
+  @Override
+  public Boolean call() {
+    out.println("___________Creating structure");
+    aab.setROOT(createDir(buildDir, "aab"));
+    if (!createStructure()) {
+      return false;
+    }
+
+    out.println("___________Extracting protobuf resources");
+    if (!extractProtobuf()) {
+      return false;
+    }
+
+    out.println("________Running bundletool");
+    if (!bundletool()) {
+      return false;
+    }
+
+    out.println("________Signing bundle");
+    if (!jarsigner()) {
+      return false;
+    }
+    return true;
+  }
+
+  private boolean createStructure() {
+    // Manifest is extracted from the protobuffed APK
+    aab.setManifestDir(createDir(aab.ROOT, "manifest"));
+
+    // Resources are extracted from the protobuffed APK
+    aab.setResDir(createDir(aab.ROOT, "res"));
+
+    // Assets are extracted from the protobuffed APK
+    aab.setAssetsDir(createDir(aab.ROOT, "assets"));
+
+    aab.setDexDir(createDir(aab.ROOT, "dex"));
+    File[] dexFiles = new File(originalDexDir).listFiles();
+    if (dexFiles != null) {
+      for (File dex : dexFiles) {
+        if (dex.isFile()) {
+          try {
+            Files.move(dex, new File(aab.dexDir, dex.getName()));
+          } catch (IOException e) {
+            e.printStackTrace();
+            return false;
+          }
+        }
+      }
+    }
+
+    aab.setLibDir(createDir(aab.ROOT, "lib"));
+    File[] libFiles = originalLibsDir.listFiles();
+    if (libFiles != null) {
+      for (File lib : libFiles) {
+        try {
+          Files.move(lib, new File(createDir(aab.ROOT, "lib"), lib.getName()));
+        } catch (IOException e) {
+          e.printStackTrace();
+          return false;
+        }
+      }
+    }
+
+    return true;
+  }
+
+  private boolean extractProtobuf() {
+    try (ZipInputStream is = new ZipInputStream(new FileInputStream(aab.getProtoApk()))) {
+      ZipEntry entry;
+      byte[] buffer = new byte[1024];
+      while ((entry = is.getNextEntry()) != null) {
+        String n = entry.getName();
+        File f = null;
+        if (n.equals("AndroidManifest.xml")) {
+          f = new File(aab.getManifestDir(), n);
+        } else if (n.equals("resources.pb")) {
+          f = new File(aab.getROOT(), n);
+        } else if (n.startsWith("assets")) {
+          f = new File(aab.getAssetsDir(), n.substring(("assets").length()));
+        } else if (n.startsWith("res")) {
+          f = new File(aab.getResDir(), n.substring(("res").length()));
+        }
+
+        if (f != null) {
+          f.getParentFile().mkdirs();
+          FileOutputStream fos = new FileOutputStream(f);
+          int len;
+          while ((len = is.read(buffer)) > 0) {
+            fos.write(buffer, 0, len);
+          }
+          fos.close();
+        }
+      }
+
+      is.close();
```

Comment:
Because try-with-resources closes the resource, the call to close here is redundant.